transports/webrtc/src/tokio/transport.rs

```diff
@@ -0,0 +1,604 @@
+    }
+}
+
+/// A stream of incoming connections on one or more interfaces.
+struct ListenStream {
+    /// The ID of this listener.
+    listener_id: ListenerId,
+
+    /// The socket address that the listening socket is bound to,
+    /// which may be a "wildcard address" like `INADDR_ANY` or `IN6ADDR_ANY`
+    /// when listening on all interfaces for IPv4 respectively IPv6 connections.
+    listen_addr: SocketAddr,
+
+    /// The config which holds this peer's certificate(s).
+    config: Config,
+
+    /// The UDP muxer that manages all ICE connections.
+    udp_mux: UDPMuxNewAddr,
+
+    /// Set to `Some` if this listener should close.
+    ///
+    /// Optionally contains a [`TransportEvent::ListenerClosed`] that should be
+    /// reported before the listener's stream is terminated.
+    report_closed: Option<Option<<Self as Stream>::Item>>,
+
+    /// Watcher for network interface changes.
+    /// Reports [`IfEvent`]s for new / deleted ip-addresses when interfaces
+    /// become or stop being available.
+    ///
+    /// `None` if the socket is only listening on a single interface.
+    if_watcher: IfWatcher,
+}
+
+impl ListenStream {
+    /// Constructs a `WebRTCListenStream` for incoming connections.
+    fn new(
+        listener_id: ListenerId,
+        config: Config,
+        udp_mux: UDPMuxNewAddr,
+        if_watcher: IfWatcher,
+    ) -> Self {
+        ListenStream {
+            listener_id,
+            listen_addr: udp_mux.listen_addr(),
+            config,
+            udp_mux,
+            report_closed: None,
+            if_watcher,
+        }
+    }
+
+    /// Report the listener as closed in a [`TransportEvent::ListenerClosed`] and
+    /// terminate the stream.
+    fn close(&mut self, reason: Result<(), Error>) {
+        match self.report_closed {
+            Some(_) => log::debug!("Listener was already closed."),
+            None => {
+                // Report the listener event as closed.
+                let _ = self
+                    .report_closed
+                    .insert(Some(TransportEvent::ListenerClosed {
+                        listener_id: self.listener_id,
+                        reason,
+                    }));
+            }
+        }
+    }
+
+    fn poll_if_watcher(&mut self, cx: &mut Context<'_>) -> Poll<<Self as Stream>::Item> {
+        while let Poll::Ready(event) = self.if_watcher.poll_if_event(cx) {
+            match event {
+                Ok(IfEvent::Up(inet)) => {
+                    let ip = inet.addr();
+                    if self.listen_addr.is_ipv4() == ip.is_ipv4()
+                        || self.listen_addr.is_ipv6() == ip.is_ipv6()
+                    {
+                        return Poll::Ready(TransportEvent::NewAddress {
+                            listener_id: self.listener_id,
+                            listen_addr: self
+                                .listen_multiaddress(ip, self.config.id_keys.public().to_peer_id()),
```

Comment:
Sorry for raising this after it has been merged, but I think that all the other transports do not include the `/p2p/` in the multiaddress they report, but this one does.

Reply:
Yeah I remember that. Does this break something? IMO, peers should always use `/p2p` when possible upon dialing another peer. We have access to the `PeerId` in the `Transport` so why not append it? Esp. because the webrtc spec mandates that you have to dial with `/p2p`.

Reply:
These addresses get sent over the network through the `identify` protocol. Either all addresses have a `/p2p/` at the end, or none, but the in-between makes no sense.

> We have access to the PeerId in the Transport so why not append it?

We already know the `PeerId` everywhere, so why should we append it to the addresses as well? It's completely redundant.

I really dislike this `/p2p/` multiaddress specifier with a passion because it's purely an API trick. It exists solely so that you can call functions with `foo(concat(address, peer_id))` (and then `foo` de-concats them) instead of `foo(address, peer_id)`. I don't understand why this `/p2p/` exists at all.

> Esp. because the webrtc spec mandates that you have to dial with /p2p.

Sorry but if the spec says that then I must say that this is completely out of scope of the spec. This is purely an API problem, not a wire problem.

Reply:
I agree that this is an API problem and would like to present a (small) counter-weight: having this syntax available is very useful so that people (in my case: IT admins) can effect a dial to a specific peer at a specific transport address by just copying&pasting a single string.

If we were to agree that that is a worthy cause for the existence of `/p2p` syntax then it would be surprising to see it rejected in API calls. `DialOpts` is already a bit weird in that sense because it requires the calling code to figure out whether the PeerId is to be required or not.

Small rant: there is quite a bit of code in `ipfs-embed` to canonicalise multiaddrs in this regard, so that the address book can present a consistent view that is useful for UIs (→ human consumption).

Reply:
I agree that it is very useful to be able to copy-paste a string that contains a `/p2p/` and pass both the address and peerid at the same time, however that is still not a proper justification for the existence of `/p2p/` in multiaddresses.

If you want your API to accept a multiaddress and peerid, then just use a type named `MultiaddrWithPeerId` and then properly split it into a `(Multiaddr, PeerId)` tuple before passing them around. Similarly, when you show them to the user, just reconstruct the `MultiaddrWithPeerId`. Not everything has to be shoved into a single `Multiaddr` type.

Reply:
True. A more sophisticated admin might want to express a relayed connection through a specific peer, in which case this separation would be difficult — in particular since multiple such hops could potentially be used (not sure whether it is supported today, but it might be!).

I’m also not sure whether I agree with the underlying premise that the PeerId is not genuinely part of the “address”. The routing example above demonstrates that an address may well depend on the identities of the nodes found at certain transport hops. This would imply that a `Multiaddr` may _contain_ a `MultiaddrWithPeerId`, at which point we’re back at square one.

Reply:
emphatic yes!